brainscore_language/models/huggingface.py

```diff
@@ -99,59 +103,97 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
-        tokens = self.tokenized_inputs['input_ids'].squeeze()
-        logits = base_output.logits.squeeze()
-        # assume that reading time is additive,
-        # i.e. reading time of multiple tokens is the sum of the perplexity of each individual token
-        cross_entropy = F.cross_entropy(logits, tokens, reduction='sum')
-        perplexity = torch.exp(cross_entropy)  # https://stackoverflow.com/a/59219379/1504411
-        return perplexity
-
-    def predict_next_word(self, base_output):
+        # `base_output.logits` is (batch_size, sequence_length, vocab_size)
+        logits = base_output.logits.squeeze(dim=0)
+
+        FIRST_TOKEN_READING_TIME = 0
+        if logits.shape[0] <= 1:  # sequence_length indicates we have only seen single word
+            # since we have no context, we can't predict the next word
+            # attempt to resolve by assuming a fixed reading time for the first word
+            return FIRST_TOKEN_READING_TIME
+
+        # get expectation (logits), shifted left by 1
+        predicted_logits = logits[-self.current_tokens['input_ids'].shape[-1] - 1: - 1, :].contiguous()
+        actual_tokens = self.current_tokens['input_ids'].squeeze(dim=0).contiguous()
+        first_token_addition = 0
+        if actual_tokens.shape[0] == predicted_logits.shape[0] + 1:  # multiple tokens for first model input
+            actual_tokens = actual_tokens[1:]  # we have no prior context to predict the 0th token
+            first_token_addition = FIRST_TOKEN_READING_TIME  # add the expected reading time for 0th token (see above)
+
+        # assume that reading time is additive, i.e. reading time of multiple tokens is
+        # the sum of the perplexity of each individual token.
+        # Note that this implementation similarly sums over the surprisal of multiple words,
+        # e.g. for the surprisal of an entire sentence.
+        cross_entropy = F.cross_entropy(predicted_logits, actual_tokens, reduction='sum') / np.log(2) \
+                        + first_token_addition
+        return cross_entropy
+
+    def predict_next_word(self, base_output: CausalLMOutput):
         """
         :param base_output: the neural network's output
         :return: predicted next word
         """
 
         logits = base_output.logits
         pred_id = torch.argmax(logits, axis=2).squeeze()
+        # Note that this is currently only predicting the next *token* which might not always be entire words.
         last_model_token_inference = pred_id[-1].tolist()
         next_word = self.tokenizer.decode(last_model_token_inference)
-
+        # `next_word` often includes a space ` ` in front of the actual word. Since the task already tells us to output
+        # a word, we can strip the space.
+        # Note that the model might also predict a token completing the current word, e.g. `fox` -> `es` (`foxes`),
+        # this is not caught in the current implementation.
+        next_word = next_word.strip()
```

Comment:
Yeah, this is slightly hacky. The right way to do this would be to have the client accept the tokens as-is and then run the model's detokenizer on output (`PreTrainedTokenizer.decode`). But if the current setup is sufficient for our current use cases, then seems good!

Reply:
agree, I think this would be good to fix going forward